packages/core/src/combinator/timeslice.js

```diff
@@ -3,111 +3,102 @@
 /** @author John Hann */
 
 import Pipe from '../sink/Pipe'
-import { disposeAll } from '@most/disposable'
+import { disposeBoth } from '@most/disposable'
 import { join } from './chain'
+import { merge } from './merge'
+import { never } from '../source/never'
+
+const timeBounds = (min, minSignal, maxSignal) =>
+  ({ min, minSignal, maxSignal })
+
+const minBounds = (minSignal) =>
+  timeBounds(Infinity, minSignal, never())
+
+const maxBounds = (maxSignal) =>
+  timeBounds(0, never(), maxSignal)
+
+const mergeTimeBounds = (b1, b2) =>
+  ({
+    min: Math.max(b1.min, b2.min),
+    minSignal: merge(b1.minSignal, b2.minSignal),
+    maxSignal: merge(b1.maxSignal, b2.maxSignal)
+  })
+
+const runTimeBounds = (bounds, timesliceSink, scheduler) =>
+  disposeBoth(
+    bounds.minSignal.run(new UpdateMinSink(timesliceSink), scheduler),
+    bounds.maxSignal.run(new UpdateMaxSink(timesliceSink), scheduler)
+  )
 
 export const until = (signal, stream) =>
-  new Until(signal, stream)
+  timeslice(maxBounds(signal), stream)
 
 export const since = (signal, stream) =>
-  new Since(signal, stream)
+  timeslice(minBounds(signal), stream)
 
 export const during = (timeWindow, stream) =>
   until(join(timeWindow), since(timeWindow, stream))
 
-class Until {
-  constructor (maxSignal, source) {
-    this.maxSignal = maxSignal
-    this.source = source
-  }
-
-  run (sink, scheduler) {
-    const min = new Bound(-Infinity, sink)
-    const max = new UpperBound(this.maxSignal, sink, scheduler)
-    const disposable = this.source.run(new TimeWindowSink(min, max, sink), scheduler)
-
-    return disposeAll([min, max, disposable])
-  }
-}
+const timeslice = (bounds, stream) =>
```

Comment:
This is the fusion algorithm.  It is essentially a right fold over adjacent `Timeslice`s, merging times bounds as it folds.